docs/design/vcpkg.md

```diff
@@ -0,0 +1,385 @@
+    * vcpkg-cmake-config:x64-linux@2024-05-23 -- git+https://github.com/microsoft/vcpkg@97a63e4bc1a17422ffe4eff71da53b4b561a7841
+  The following packages will be rebuilt:
+      fmt:x64-linux@10.1.1 -- git+https://github.com/microsoft/vcpkg@dfe9aa860f5a8317f341a21d317be1cf44e89f18
+  The following packages will be built and installed:
+      zlib:x64-linux@1.3.1 -- git+https://github.com/microsoft/vcpkg@3f05e04b9aededb96786a911a16193cdb711f0c9
+  ```
+
+</details>
+
+While it contains enough data to populate a SBOM it has to be parsed which is
+inherently error-prone. There does not seem to be any way to get around this.
+
+Another bigger problem is the fact that `vcpkg` runs cmake scripts. For the
+bulk of the ports this is not a problem since these scripts are very
+straightforward and boil down to a number of simple actions like downloading
+packages and extracting archives. However for more complex ports it is hard to
+tell at a glance what exactly is happening during installation: some define
+large libraries of auxiliary cmake functions, others partially parse downloaded
+sources and make decisions basing on parsing results. While there is some degree
+of trust to ports found in the default registry other registries should be
+treated with suspicion. It is worth noting that cmake allows execution of
+external programs and while vcpkg makes efforts to limit which programs can
+and cannot be executed (simple brute-force tests turned out negative)
+it is impossible to say with certainty that a malicious program cannot
+be injected and executed this way.
+
+
+### Option 2
+
+Another possible way of handling vcpkg dependencies is to implement the
+relevant parts of vcpkg ourselves. The relevant parts are `vcpkg.json`
+processing and interpreting those commands in portfiles which are necessary for
+downloading sources.  This would require the following:
+
+ 1. constructing dependencies tree for a package ourselves;
+ 2. consulting a registry to collect relevant ports;
+ 3. converting a git registry into a filesystem registry;
+ 4. downloading the sources;
+ 5. amending the project to use the new filesystem registry.
+
+After these steps it should be possible to build the project hermetically.
+
+`1.` is relatively straightforward on the surface and requires traversing jsons
+in a registry.  Once all dependencies are known `ports` directory must be
+traversed to extract data about versions (the latest available unless
+explicitly specified otherwise) and upstream location.
+
+`2. ` requires some rudimentary parsing of cmake scripts. [cmake_parser][]
+appears to be a good enough readily available solution for that given that
+cmake does not publish the language grammar. An alternative would be to produce
+either an ad-hoc parser with [PyParsing][] or to invest into cmake grammar
+specification and then use some standard parser generator. Preliminary
+experiments have shown that `cmake_parser` is good enough for the task.
```

Comment:
Hmm, a plausible approach to use the python package. That said, it looks like it's one of them abandonware, so we need to keep that in mind. Out of curiosity, have you tried to write a cmake native script that would try to parse the data? I think that might be a worth AI exercise these days and see what it spits out :) .

Reply:
Possible, but I would rather not touch cmake unless strictly necessary. Even if all parsers, parser generators and parser libraries go extinct writing one for cmake from scratch would not be an excessively complicated task, the language itself is rather primitive.